### PR TITLE
[SPARK-19826][ML][PYTHON]add spark.ml Python API for PIC

### PR DESCRIPTION
**Pull request by huaxingao**
## What changes were proposed in this pull request?

add spark.ml Python API for PIC

## How was this patch tested?

add doctest

### CODE REVIEW

**Comment by huaxingao** [comment]
@jkbradley Could you please review when you have time? Thank you very much in advance!

File: python/pyspark/ml/clustering.py [hunk 1]
```diff
@@ -1156,6 +1156,201 @@ def getKeepLastCheckpoint(self):
         return self.getOrDefault(self.keepLastCheckpoint)
 
 
+class _PowerIterationClusteringParams(JavaParams, HasMaxIter, HasPredictionCol):
```
Comment (WeichenXu123) [inline python/pyspark/ml/clustering.py]: Why not directly add params into class `PowerIterationClustering`?

File: python/pyspark/ml/clustering.py [hunk 2]
```diff
@@ -1156,6 +1156,201 @@ def getKeepLastCheckpoint(self):
...
+        Gets the value of `similaritiesCol`
+        """
+        return self.getOrDefault(self.binary)
+
+
+@inherit_doc
+class PowerIterationClustering(JavaTransformer, _PowerIterationClusteringParams, JavaMLReadable,
+                               JavaMLWritable):
+    """
+    Model produced by [[PowerIterationClustering]].
```
Comment (WeichenXu123) [inline python/pyspark/ml/clustering.py]: The doc is wrong. Copy doc from scala side.

File: python/pyspark/ml/clustering.py [hunk 3]
```diff
@@ -1156,6 +1156,201 @@ def getKeepLastCheckpoint(self):
...
+        kwargs = self._input_kwargs
+        self.setParams(**kwargs)
+
+    @keyword_only
+    @since("2.4.0")
+    def setParams(self, predictionCol="prediction", k=2, maxIter=20, initMode="random",
+                  idCol="id", neighborsCol="neighbors", similaritiesCol="similarities"):
+        """
+        setParams(self, predictionCol="prediction", k=2, maxIter=20, initMode="random",\
+                  idCol="id", neighborsCol="neighbors", similaritiesCol="similarities"):
```
Comment (WeichenXu123) [inline python/pyspark/ml/clustering.py]: remove `:` at the end.

File: python/pyspark/ml/clustering.py [hunk 4]
```diff
@@ -1156,6 +1156,201 @@ def getKeepLastCheckpoint(self):
...
+    'degree'
+
+    .. versionadded:: 2.4.0
+    """
+    @keyword_only
+    def __init__(self, predictionCol="prediction", k=2, maxIter=20, initMode="random",
+                 idCol="id", neighborsCol="neighbors", similaritiesCol="similarities"):
+        """
+        __init__(self, predictionCol="prediction", k=2, maxIter=20, initMode="random",\
+                 idCol="id", neighborsCol="neighbors", similaritiesCol="similarities"):
```
Comment (WeichenXu123) [inline python/pyspark/ml/clustering.py]: remove `:` at the end.

File: python/pyspark/ml/clustering.py [hunk 5]
```diff
@@ -1156,6 +1156,201 @@ def getKeepLastCheckpoint(self):
...
+    similaritiesCol = Param(Params._dummy(), "similaritiesCol",
+                            "Name of the input column for non-negative weights (similarities) " +
+                            "of edges between the vertex in `idCol` and each neighbor in " +
+                            "`neighborsCol`",
+                            typeConverter=TypeConverters.toString)
+
+    @since("2.4.0")
+    def getK(self):
+        """
+        Gets the value of `k`
```
Comment (WeichenXu123) [inline python/pyspark/ml/clustering.py]: Should use:
:py:attr:`k`
and update everywhere else.

File: python/pyspark/ml/clustering.py [hunk 6]
```diff
@@ -1156,6 +1156,201 @@ def getKeepLastCheckpoint(self):
...
+    >>> n = n1 + n2
+    >>> points = genCircle(r1, n1) + genCircle(r2, n2)
+    >>> similarities = []
+    >>> for i in range (1, n):
+    ...    neighbor = []
+    ...    weight = []
+    ...    for j in range (i):
+    ...        neighbor.append((long)(j))
+    ...        weight.append(sim(points[i], points[j]))
+    ...    similarities.append([(long)(i), neighbor, weight])
```
Comment (WeichenXu123) [inline python/pyspark/ml/clustering.py]: The doctest code looks like too long, maybe more proper to put it in examples.
Could you replace the data generation code here by using a simple hardcoded dataset ?

**Review by WeichenXu123 — COMMENTED** [review]
Thanks! I made a pass.

File: python/pyspark/ml/clustering.py (hunk 1 above)
Comment (huaxingao) [reply python/pyspark/ml/clustering.py]: @WeichenXu123 Thanks for your review. The params can be either inside class PowerIterationClustering or separate. I will move them back inside class PowerIterationClustering, to be consistent with the params in the other classes in clustering.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/PowerIterationClustering.scala [hunk 7]
```diff
@@ -97,13 +97,15 @@ private[clustering] trait PowerIterationClusteringParams extends Params with Has
   def getNeighborsCol: String = $(neighborsCol)
 
   /**
-   * Param for the name of the input column for neighbors in the adjacency list representation.
+   * Param for the name of the input column for non-negative weights (similarities) of edges
+   * between the vertex in `idCol` and each neighbor in `neighborsCol`.
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/PowerIterationClustering.scala]: Good catch!

File: python/pyspark/ml/clustering.py [hunk 8]
```diff
@@ -1156,6 +1156,204 @@ def getKeepLastCheckpoint(self):
...
+    def getIdCol(self):
+        """
+        Gets the value of :py:attr:`idCol`.
+        """
+        return self.getOrDefault(self.idCol)
+
+    @since("2.4.0")
+    def setNeighborsCol(self, value):
+        """
+        Sets the value of :py:attr:`neighborsCol.
```
Comment (viirya) [inline python/pyspark/ml/clustering.py]: Missing the left back quote.

File: python/pyspark/ml/clustering.py [hunk 9]
```diff
@@ -1156,6 +1156,204 @@ def getKeepLastCheckpoint(self):
...
+        Sets the value of :py:attr:`similaritiesCol`.
+        """
+        return self._set(similaritiesCol=value)
+
+    @since("2.4.0")
+    def getSimilaritiesCol(self):
+        """
+        Gets the value of :py:attr:`similaritiesCol`.
+        """
+        return self.getOrDefault(self.binary)
```
Comment (viirya) [inline python/pyspark/ml/clustering.py]: `self.binary` -> `self.similaritiesCol`?

File: python/pyspark/ml/clustering.py [hunk 10]
```diff
@@ -1156,6 +1156,204 @@ def getKeepLastCheckpoint(self):
...
+    <a href=http://www.icml2010.org/papers/387.pdf>Lin and Cohen</a>. From the abstract:
+    PIC finds a very low-dimensional embedding of a dataset using truncated power
+    iteration on a normalized pair-wise similarity matrix of the data.
+
+    PIC takes an affinity matrix between items (or vertices) as input.  An affinity matrix
+    is a symmetric matrix whose entries are non-negative similarities between items.
+    PIC takes this matrix (or graph) as an adjacency matrix.  Specifically, each input row
+    includes:
+
+     - :py:class:`idCol`: vertex ID
```
Comment (viirya) [inline python/pyspark/ml/clustering.py]: ```:py:attr:`idCol` ```? And also the below ```:py:class:`neighborsCol` ```, etc...

File: python/pyspark/ml/clustering.py [hunk 11]
```diff
@@ -1156,6 +1156,204 @@ def getKeepLastCheckpoint(self):
...
+    each row (vertex).
+
+    Notes:
+
+     - [[PowerIterationClustering]] is a transformer with an expensive [[transform]] operation.
+        Transform runs the iterative PIC algorithm to cluster the whole input dataset.
+     - Input validation: This validates that similarities are non-negative but does NOT validate
+        that the input matrix is symmetric.
+
+    @see <a href=http://en.wikipedia.org/wiki/Spectral_clustering>
```
Comment (viirya) [inline python/pyspark/ml/clustering.py]: Use `.. seealso::`?

File: python/pyspark/ml/clustering.py [hunk 12]
```diff
@@ -1156,6 +1156,204 @@ def getKeepLastCheckpoint(self):
...
+     - :py:class:`idCol`: vertex ID
+     - :py:class:`neighborsCol`: neighbors of vertex in :py:class:`idCol`
+     - :py:class:`similaritiesCol`: non-negative weights (similarities) of edges between the
+        vertex in :py:class:`idCol` and each neighbor in :py:class:`neighborsCol`
+
+    PIC returns a cluster assignment for each input vertex.  It appends a new column
+    :py:class:`predictionCol` containing the cluster assignment in :py:class:`[0,k)` for
+    each row (vertex).
+
+    Notes:
```
Comment (viirya) [inline python/pyspark/ml/clustering.py]: Use `.. note::`?

File: python/pyspark/ml/clustering.py [hunk 13]
```diff
@@ -1156,6 +1156,205 @@ def getKeepLastCheckpoint(self):
...
+
+    .. note::
+
+     - [[PowerIterationClustering]] is a transformer with an expensive [[transform]] operation.
+        Transform runs the iterative PIC algorithm to cluster the whole input dataset.
+     - Input validation: This validates that similarities are non-negative but does NOT validate
+        that the input matrix is symmetric.
+
+    .. seealso:: <a href=http://en.wikipedia.org/wiki/Spectral_clustering>
+    Spectral clustering (Wikipedia)</a>
```
Comment (viirya) [inline python/pyspark/ml/clustering.py]: You can check other places using `seealso`:

```python
.. seealso:: `Spectral clustering \
<http://en.wikipedia.org/wiki/Spectral_clustering>`_
```

**Comment by jkbradley** [comment]
I think we messed up with the original PIC API.  Could you please check out my comment here https://issues.apache.org/jira/browse/SPARK-15784 ?  If others agree, I'll revert the Scala API and we can work on adding a modified version.

**Comment by huaxingao** [comment]
@jkbradley 
Thanks for letting me know. I will change the python API accordingly after the new scala version is in.

**Comment by mengxr** [comment]
@huaxingao We updated the Scala/Java API in https://github.com/apache/spark/pull/21493. Could you update this PR for the Python API? It should be similar to the PrefixSpan Python API (https://github.com/apache/spark/commit/90ae98d1accb3e4b7d381de072257bdece8dd7e0), which is neither a transformer nor an estimator. Let me know if you don't have time. @WeichenXu123  could update the Python API as well.

**Comment by huaxingao** [comment]
@mengxr @WeichenXu123 I will update this. Thanks.

**Comment by mengxr** [comment]
@huaxingao Any updates?

**Comment by huaxingao** [comment]
@mengxr Sorry for the delay. I will submit an update later today. Do you want me to close this PR and do a new one? or just update this PR?

**Comment by huaxingao** [comment]
@mengxr @WeichenXu123 I will close this one and submit a new PR soon. Thanks!

**Pull request closed without merging**